### PR TITLE
Issues/2368 registry auth

### LINKED ISSUE #2368: Apply authorisation to dereferenced records when requested via /record/id API
### Problem description
Building on #2364, this applies authorisation to aspects retrieved through links when `dereference=true` is specified through the [/record/{id}](https://dev.magda.io/api/v0/apidocs/#api-Registry_Record_Service-GetV0RegistryRecordsId) endpoint in the registry.

### Acceptance Criteria
(assuming `dereference=true` is specified)
- If a user is allowed to see a requested record aspect, and that aspect links to another record  that the user is allowed to see, the inner aspect should be returned as part of the outer one, just as it was before
- If the user is allowed to see an outer aspect but not the inner aspect it links to, there should be no inner aspect present - as if the inner record aspect row didn't exist. **(e.g. if the dataset-distributions schema links to a single distribution record but the user isn't allowed to see that distribution, the `distributions` property should be `[]` but you should still be able to see the dataset-distributions aspect.**
- If the user is not allowed to see the outer aspect but is allowed to see the inner one, they still shouldn't be able to see either when dereferencing.

**(with `deference=false`)**
  - **The id of dereferenced records that the user doesn't have access to should be suppressed... e.g. in the example above where a user has access to a dataset record but not the distribution, the `distributions` property should also be `[]`. **

### Technical Notes
- This should be achievable using similar logic to getting records without dereferencing. Dereferencing works by examining the aspect definition, pulling out linked aspects and then querying for them. When the linked aspects are queries, it should be possible to look up the possible policies from OPA, turn them into SQL and include them in the query for the inner aspect.

### Out of Scope
- Aspect-level authorisation

### PR DESCRIPTION
**Pull request by mwu2018**
### Add authz to record linking aspects

Fixes #2368 
1. For some historical reason, many authz related issues still imply that authz is applied at aspect level. However, this PR only applies authz at record level, including records referenced via a linking aspect (by either single link or link array). Therefore all authz related issues should be revised to reflect this fact.
2. If a record aspect references an empty link array, or a user is not authorised to access any records in the link array, a user will get an empty array if querying for that link array aspect with dereference param set to either true or false. This behaviour is specified by the criteria. 
However, in the case that 
  -- a record references another record via a single link aspect and 
  -- a user queries for that single link aspect and 
  -- the user is authorised to access the referencing record but not the referenced record,
he will get different results for different dereference param settings.
  -- When dereference param set to false, the result is an empty string "" as record ID of the referenced record.
  -- When dereference parma set to true, the result is an empty JsObject {}.
3. With this PR, all registry read operations (except for history) will have access control.
4. All comments raised in PR 2441 are addressed.
5. The specifications are verified by tests opa/RecordsOpaSpec.

This PR is built on top of the PR https://github.com/magda-io/magda/pull/2411.

### Checklist

-   [x] There are unit tests to verify my changes are correct or unit tests aren't applicable
-   [x] I've updated CHANGES.md with what I changed.
-   [x] I've linked this PR to an issue in ZenHub (core dev team only)

### CODE REVIEW

**Comment by AlexGilleran** [comment]
Good stuff :) I still haven't given it a test but will shortly

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala [hunk 1]
```diff
@@ -1174,29 +1182,44 @@ object DefaultRecordPersistence
...
-      dereferenceDetails
+      opaQuery,
+      dereference.getOrElse(false)
     )
 
     val limit = rawLimit
       .map(l => Math.min(l, maxResultCount))
       .getOrElse(defaultResultCount)
 
+    val tempName = SQLSyntax.createUnsafely("temp")
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: done

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala [hunk 2]
```diff
@@ -1303,32 +1331,35 @@ object DefaultRecordPersistence
...
+      val aspects: List[(String, JsObject)] =
         sql"""select aspectId, jsonSchema
             from Aspects
-            where aspectId in ($aspectIds)"""
-          .map(
-            rs =>
+            where aspectId in ($aspectIds)
+          """
+          .map(rs => {
+            if (rs.string("jsonSchema") != null)
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: done

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala [hunk 3]
```diff
@@ -1303,32 +1331,35 @@ object DefaultRecordPersistence
...
               (
                 rs.string("aspectId"),
                 JsonParser(rs.string("jsonSchema")).asJsObject
               )
-          )
+            else
+              (rs.string("aspectId"), null)
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: done

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala [hunk 4]
```diff
@@ -1408,61 +1439,130 @@ object DefaultRecordPersistence
...
+  private def opaQueriesToWhereParts(
+      tenantId: BigInt,
+      opaQueries: Seq[OpaQuery],
+      operationType: AuthOperations.OperationType
+  ): Iterable[SQLSyntax] = {
+    val theOpaQueries: List[OpaQuery] =
+      if (opaQueries.nonEmpty) opaQueries.toList
+      else List(OpaQuerySkipAccessControl)
+    val opaSql: SQLSyntax =
+      sqls"(${SQLSyntax.joinWithAnd(
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: removed

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala [hunk 5]
```diff
@@ -1408,61 +1439,130 @@ object DefaultRecordPersistence
...
+    val result = List(opaSql)
+    result
+  }
+
+  private def getAspectColumnName(
+      index: Int,
+      prefix: String = ""
+  ): SQLSyntax = {
+    // Use a simple numbered column name rather than trying to make the aspect name safe.
+    SQLSyntax.createUnsafely(prefix + s"aspect$index")
```
[the remark this answers is not in the document]
Comment (mwu2018) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: done

**Comment by mwu2018** [comment]
"If the user doesn't have access to a single distribution of a dataset then they can't see any distributions on that dataset through the Dataset page."
Regarding this issue, a test case shows the opposite:  record-5 (dataset) links to record-1 and record-3 (distributions). userId2 has access to dataset record-5 and gets distributions that include record-3 but not record-1.

File: magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala [hunk 6]
```diff
@@ -1408,61 +1440,136 @@ object DefaultRecordPersistence
...
+      dereference: Boolean = false
   ): Iterable[scalikejdbc.SQLSyntax] = {
+    val opaConditions = getOpaConditions(tenantId, opaQueries, operationType)
+
     val result: Iterable[SQLSyntax] = aspectIds.zipWithIndex.map {
       case (aspectId, index) =>
-        // Use a simple numbered column name rather than trying to make the aspect name safe.
-        val aspectColumnName = SQLSyntax.createUnsafely(s"aspect$index")
-        val selection = dereferenceDetails
+        val referenceDetails =
```
Comment (AlexGilleran) [inline magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: Could this be out of the loop (like on line 1490?)

Comment (mwu2018) [reply magda-registry-api/src/main/scala/au/csiro/data61/magda/registry/RecordPersistence.scala]: Yes. It was silly to be put in the loop.

**Comment by AlexGilleran** [comment]
Apologies @mwu2018, I didn't actually test that exact case and made an assumption, my bad :). That scenario works as expected, but the other one doesn't.

What I tested was this:

I've made a record "magda-dist-057ce084-d02f-4bb7-a526-3c7fb723d08c" on https://issues-2368-registry-auth.dev.magda.io.

If the `dataset-distributions` aspect looks like

```json
{
    "distributions": []
}
```

Then it behaves as expected - https://issues-2368-registry-auth.dev.magda.io/api/v0/registry/records/magda-ds-e062b1ea-29dc-4bd6-9fd5-3652d578701a?aspect=dataset-distributions returns

```json
{
  "aspects": {
    "dataset-distributions": {
      "distributions": []
    }
  },
  "id": "magda-ds-e062b1ea-29dc-4bd6-9fd5-3652d578701a",
  "name": "Magda Marketecture",
  "tenantId": 0
}
```

However if I add a distribution that I don't have access to:

```
{
    "distributions": ["magda-dist-057ce084-d02f-4bb7-a526-3c7fb723d08c"]
}
```

Then https://issues-2368-registry-auth.dev.magda.io/api/v0/registry/records/magda-ds-e062b1ea-29dc-4bd6-9fd5-3652d578701a?aspect=dataset-distributions returns `404`... but if it were behaving as if the distribution didn't exist then it would return the same as the first time.

**Comment by mwu2018** [comment]
When I wrote the test case "and not return any links referenced by record-4 to anonymous user (dereference=true)", it returns 404. I thought that that was expected. I can modify it to return 200 with empty array.

**Comment by AlexGilleran** [comment]
Great :)

**Pull request merged**